### PR TITLE
more opcodes

### PR DESCRIPTION
**Pull request by Valgard**

### CODE REVIEW

File: src/emulator/cpu.ts [hunk 1]
```diff
@@ -109,26 +109,76 @@ export class Cpu {
...
     private dispatch(instruction: Instruction): number {
         if (instruction.op !== Operation.invalid) this.onExecute.dispatch(this.state.p);
 
         switch (instruction.op) {
+            case Operation.adc: {
+                this.clock.increment(instruction.cycles);
+
+                const a = this.state.r8[r8.a];
+                const operand = this.getArg1(instruction);
+                const result = a + operand + (this.state.r8[r8.f] & flag.c);
```
Comment (DirtyHairy) [inline src/emulator/cpu.ts]: I am pretty sure this should be  `(this.state.r8[r8.f] & flag.c) >>> 4` --- carry is bit 4. This is usually used for 16 bit additions by first adding the low bytes and then adding the high bytes with carry, so an overflow on the low bytes amounts to `carry high` amounts to an additional contribution of 1 to the high bytes.

File: src/emulator/cpu.ts [hunk 2]
```diff
@@ -340,19 +383,56 @@ export class Cpu {
...
 
                 return instruction.cycles;
             }
 
+            case Operation.sbc: {
+                this.clock.increment(instruction.cycles);
+
+                const a = this.state.r8[r8.a];
+                const operand = this.getArg1(instruction);
+                const result = a - operand - (this.state.r8[r8.f] & flag.c);
```
Comment (DirtyHairy) [inline src/emulator/cpu.ts]: Same as `adc` --- this should be `(this.state.r8[r8.f] & flag.c) >>> 4`

File: src/emulator/cpu.ts [hunk 3]
```diff
@@ -109,26 +109,76 @@ export class Cpu {
...
-        this.bus.write(this.state.r16[r16.sp], this.state.p & 0xff);
+        this.stackPush(this.state.p);
 
         this.state.p = getIrqVector(irq);
         this.clock.increment(5);
 
         return 5;
     }
 
+    private stackPush(value: number): void {
```
Comment (DirtyHairy) [inline src/emulator/cpu.ts:120]: Maybe rename that to `stackPush16` for clarity? Although there is no 8 bit version on the gameboy, so 🤷‍♂️

File: src/emulator/cpu.ts [hunk 4]
```diff
@@ -109,26 +109,76 @@ export class Cpu {
...
+        value = value & 0xffff;
+        this.state.r16[r16.sp] = (this.state.r16[r16.sp] - 0x01) & 0xffff;
+        this.bus.write(this.state.r16[r16.sp], value >>> 8);
+        this.state.r16[r16.sp] = (this.state.r16[r16.sp] - 0x01) & 0xffff;
+        this.bus.write(this.state.r16[r16.sp], value & 0xff);
+    }
+
+    private stackPop(): number {
```
Comment (DirtyHairy) [inline src/emulator/cpu.ts:128]: Maybe rename that to `stackPop16` for clarity? Although there is no 8 bit version on the gameboy, so 🤷‍♂️

File: src/emulator/cpu.ts [hunk 5]
```diff
@@ -109,26 +109,76 @@ export class Cpu {
...
+                const a = this.state.r8[r8.a];
+                const operand = this.getArg1(instruction);
+                const result = a + operand + (this.state.r8[r8.f] & flag.c);
+
+                this.state.r8[r8.a] = result & 0xffff;
+
+                // prettier-ignore
+                this.state.r8[r8.f] =
+                    (result === 0 ? flag.z : 0x00) |
+                    ((((a & 0xf) + (operand & 0xf) + (this.state.r8[r8.f] & flag.c)) > 0xf) ? flag.h : 0x00) |
```
Comment (DirtyHairy) [inline src/emulator/cpu.ts]: See above.

File: src/emulator/cpu.ts [hunk 6]
```diff
@@ -340,19 +383,56 @@ export class Cpu {
...
+                const a = this.state.r8[r8.a];
+                const operand = this.getArg1(instruction);
+                const result = a - operand - (this.state.r8[r8.f] & flag.c);
+
+                this.state.r8[r8.a] = result & 0xffff;
+
+                // prettier-ignore
+                this.state.r8[r8.f] =
+                    (result === 0 ? flag.z : 0x00) |
+                    ((((a & 0xf) - (operand & 0xf) - (this.state.r8[r8.f] & flag.c)) < 0) ? flag.h : 0x00) |
```
Comment (DirtyHairy) [inline src/emulator/cpu.ts]: See above.

**Review by DirtyHairy — CHANGES_REQUESTED** [review]
Nice work! I have a few comments and corrections in `cpu.ts`.

What happened to `yarn.lock`? Seems like configuration foo between you side and mine --- the registry changed from yarn to npm 😏 If this wasn't intentional I would like to leave it out of the merge.

**Comment by Valgard** [comment]
> 

was an accident, fixed ;)

**Pull request merged**